core/blockchain.go

```diff
@@ -265,48 +269,53 @@ func NewBlockChain(db ethdb.Database, cacheConfig *CacheConfig, chainConfig *par
 			txIndexBlock = frozen
 		}
 	}
-
 	if err := bc.loadLastState(); err != nil {
 		return nil, err
 	}
+	// Make sure the state associated with the block is available
+	head := bc.CurrentBlock()
+	if _, err := state.New(head.Root(), bc.stateCache, bc.snaps); err != nil {
+		log.Warn("Head state missing, repairing", "number", head.Number(), "hash", head.Hash())
+		if err := bc.SetHead(head.NumberU64()); err != nil {
+			return nil, err
+		}
+	}
 	// The first thing the node will do is reconstruct the verification data for
 	// the head block (ethash cache or clique voting snapshot). Might as well do
 	// it in advance.
 	bc.engine.VerifyHeader(bc, bc.CurrentHeader(), true)
-
-	if frozen, err := bc.db.Ancients(); err == nil && frozen > 0 {
-		var (
-			needRewind bool
-			low        uint64
-		)
-		// The head full block may be rolled back to a very low height due to
-		// blockchain repair. If the head full block is even lower than the ancient
-		// chain, truncate the ancient store.
-		fullBlock := bc.CurrentBlock()
-		if fullBlock != nil && fullBlock != bc.genesisBlock && fullBlock.NumberU64() < frozen-1 {
-			needRewind = true
-			low = fullBlock.NumberU64()
-		}
-		// In fast sync, it may happen that ancient data has been written to the
-		// ancient store, but the LastFastBlock has not been updated, truncate the
-		// extra data here.
-		fastBlock := bc.CurrentFastBlock()
-		if fastBlock != nil && fastBlock.NumberU64() < frozen-1 {
-			needRewind = true
-			if fastBlock.NumberU64() < low || low == 0 {
-				low = fastBlock.NumberU64()
+	/*
+		// Ensure that the head block is above the freezer limit
+		if frozen, err := bc.db.Ancients(); err == nil && frozen > 0 {
```

Comment:
Why not just remove all this?

Reply:
Whoops, I commented out that part because I figured it's not needed, then forgot about it. I'll need to double check this.

Reply:
And also in `SetHead` we update the markers first and the do the deletion work. If panic happens in the between, we need to heal the chain.

Reply:
Ok, I deleted it altogether. The purpose of this code was that after a crash-restart-repair, if the repair moves the markers before the freezer limit, then the freezer should be hotfixed up too. This special casing is not needed any more because all code paths use SetHead, and this unified method will ensure that the chain ends up in a consistent state, independent of why sethead was called.

Reply:
I've reverted it btw.